File: scripts/generate_calendar.py
#!/usr/bin/env python3
"""Generate an iCalendar file from a GitHub Projects v2 calendar."""

from __future__ import annotations

import argparse
import json
import os
import sys
import urllib.error
import urllib.request
from dataclasses import dataclass
from datetime import date, datetime, timedelta, timezone
from pathlib import Path
from typing import Any, Iterable

API_URL = "https://api.github.com/graphql"

QUERY = """
query($organization: String!, $project: Int!, $cursor: String) {
  organization(login: $organization) {
    projectV2(number: $project) {
      title
      items(first: 100, after: $cursor) {
        nodes {
          updatedAt
          content {
            ... on Issue {
              number
              title
              url
              updatedAt
              repository { nameWithOwner }
            }
          }
          fieldValues(first: 50) {
            nodes {
              ... on ProjectV2ItemFieldDateValue {
                date
                field { ... on ProjectV2FieldCommon { name } }
              }
            }
          }
        }
        pageInfo { hasNextPage endCursor }
      }
    }
  }
}
"""


@dataclass(frozen=True)
class Event:
    uid: str
    title: str
    url: str
    start: date
    end: date  # Inclusive.
    updated_at: datetime


def graphql(token: str, variables: dict[str, Any]) -> dict[str, Any]:
    request = urllib.request.Request(
        API_URL,
        data=json.dumps({"query": QUERY, "variables": variables}).encode(),
        headers={
            "Accept": "application/vnd.github+json",
            "Authorization": f"Bearer {token}",
            "Content-Type": "application/json",
            "User-Agent": "zunoser-calendar",
            "X-GitHub-Api-Version": "2022-11-28",
        },
    )
    try:
        with urllib.request.urlopen(request, timeout=30) as response:
            result = json.load(response)
    except urllib.error.HTTPError as error:
        detail = error.read().decode(errors="replace")
        raise RuntimeError(
            f"GitHub API returned HTTP {error.code}: {detail}"
        ) from error
    except urllib.error.URLError as error:
        raise RuntimeError(f"GitHub API request failed: {error.reason}") from error

    if result.get("errors"):
        messages = "; ".join(
            error.get("message", "unknown error") for error in result["errors"]
        )
        raise RuntimeError(f"GitHub GraphQL error: {messages}")
    return result["data"]


def fetch_items(
    token: str, organization: str, project: int
) -> tuple[str, list[dict[str, Any]]]:
    cursor = None
    items: list[dict[str, Any]] = []
    title = "Calendar"

    while True:
        data = graphql(
            token,
            {"organization": organization, "project": project, "cursor": cursor},
        )
        project_data = data.get("organization", {}).get("projectV2")
        if project_data is None:
            raise RuntimeError(
                f"Project {organization}#{project} was not found or is not accessible"
            )
        title = project_data["title"]
        connection = project_data["items"]
        items.extend(connection["nodes"])
        page_info = connection["pageInfo"]
        if not page_info["hasNextPage"]:
            return title, items
        cursor = page_info["endCursor"]


def parse_date(value: Any) -> date | None:
    if not isinstance(value, str):
        return None
    try:
        parsed = date.fromisoformat(value)
    except ValueError:
        return None
    return parsed if value == parsed.isoformat() else None


def event_from_item(
    item: dict[str, Any], repository: str, start_field: str, end_field: str
) -> Event | None:
    issue = item.get("content")
    if not issue or issue.get("repository", {}).get("nameWithOwner") != repository:
        return None

    fields = {
        node.get("field", {}).get("name"): node.get("date")
        for node in item.get("fieldValues", {}).get("nodes", [])
        if node.get("field")
    }
    start = parse_date(fields.get(start_field))
    end = parse_date(fields.get(end_field))
    if start is None and end is None:
        return None
    if start is None:
        start = end
    if end is None:
        end = start
    assert start is not None and end is not None
    if end < start:
        print(
            f"warning: skipped {issue['url']}: {end_field} precedes {start_field}",
            file=sys.stderr,
        )
        return None

    updated_at = datetime.fromisoformat(
        item.get("updatedAt", issue["updatedAt"]).replace("Z", "+00:00")
    )
    return Event(
        f"{repository}#{issue['number']}",
        issue["title"],
        issue["url"],
        start,
        end,
        updated_at,
    )


def escape_text(value: str) -> str:
    return (
        value.replace("\\", "\\\\")
        .replace(";", "\\;")
        .replace(",", "\\,")
        .replace("\r\n", "\\n")
        .replace("\n", "\\n")
        .replace("\r", "\\n")
    )


def fold_line(line: str) -> list[str]:
    """Fold an iCalendar content line to at most 75 UTF-8 octets."""
    parts: list[str] = []
    current = ""
    limit = 75
    for character in line:
        if len((current + character).encode()) > limit:
            parts.append(current)
            current = " " + character
            limit = 75
        else:
            current += character
    parts.append(current)
    return parts


def render_calendar(name: str, events: Iterable[Event]) -> bytes:
    lines = [
        "BEGIN:VCALENDAR",
        "VERSION:2.0",
        "PRODID:-//zunoser//GitHub Project Calendar//EN",
        "CALSCALE:GREGORIAN",
        "METHOD:PUBLISH",
        f"X-WR-CALNAME:{escape_text(name)}",
    ]
    for event in sorted(events, key=lambda value: (value.start, value.end, value.uid)):
        timestamp = event.updated_at.astimezone(timezone.utc).strftime(
            "%Y%m%dT%H%M%SZ"
        )
        lines.extend(
            [
                "BEGIN:VEVENT",
                f"UID:{event.uid}@github.com",
                f"DTSTAMP:{timestamp}",
                f"LAST-MODIFIED:{timestamp}",
                f"DTSTART;VALUE=DATE:{event.start:%Y%m%d}",
                f"DTEND;VALUE=DATE:{event.end + timedelta(days=1):%Y%m%d}",
                f"SUMMARY:{escape_text(event.title)}",
                f"URL:{event.url}",
                "END:VEVENT",
            ]
        )
    lines.append("END:VCALENDAR")
    content = "\r\n".join(part for line in lines for part in fold_line(line))
    return (content + "\r\n").encode()


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--organization", default="zunoser")
    parser.add_argument("--project", type=int, default=3)
    parser.add_argument("--repository", default="zunoser/calendar")
    parser.add_argument("--start-field", default="Start Date")
    parser.add_argument("--end-field", default="Target Date")
    parser.add_argument("--output", type=Path, default=Path("calendar.ics"))
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    token = os.environ.get("PROJECT_TOKEN")
    if not token:
        print("PROJECT_TOKEN is required", file=sys.stderr)
        return 2
    try:
        name, items = fetch_items(token, args.organization, args.project)
        events = [
            event
            for item in items
            if (
                event := event_from_item(
                    item, args.repository, args.start_field, args.end_field
                )
            )
            is not None
        ]
        content = render_calendar(name, events)
        args.output.write_bytes(content)
    except (OSError, RuntimeError, ValueError, KeyError, TypeError) as error:
        print(f"error: {error}", file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: src/generate-calendar.ts
import { Temporal } from "@js-temporal/polyfill";
import { graphql } from "@octokit/graphql";
import ical, { ICalCalendarMethod } from "ical-generator";
import { z } from "zod";

const ORGANIZATION = "zunoser";
const PROJECT_NUMBER = 3;
const REPOSITORY = "zunoser/calendar";
const START_FIELD = "Start Date";
const END_FIELD = "Target Date";
const PAGE_SIZE = 100;

const query = `
  query CalendarProject(
    $organization: String!
    $project: Int!
    $startField: String!
    $endField: String!
    $cursor: String
  ) {
    organization(login: $organization) {
      projectV2(number: $project) {
        title
        items(first: ${PAGE_SIZE}, after: $cursor) {
          totalCount
          nodes {
            updatedAt
            content {
              __typename
              ... on Issue {
                number
                title
                url
                repository { nameWithOwner }
              }
            }
            startDate: fieldValueByName(name: $startField) {
              ... on ProjectV2ItemFieldDateValue { date }
            }
            endDate: fieldValueByName(name: $endField) {
              ... on ProjectV2ItemFieldDateValue { date }
            }
          }
          pageInfo { hasNextPage endCursor }
        }
      }
    }
  }
`;

const issueSchema = z.object({
  __typename: z.literal("Issue"),
  number: z.int().positive(),
  title: z.string(),
  url: z.url(),
  repository: z.object({ nameWithOwner: z.string() }),
});

const nonIssueSchema = z.object({
  __typename: z.enum(["DraftIssue", "PullRequest"]),
});

const itemSchema = z.object({
  updatedAt: z.iso.datetime({ offset: true }),
  content: z.discriminatedUnion("__typename", [issueSchema, nonIssueSchema]).nullable(),
  startDate: z.object({ date: z.unknown() }).nullable(),
  endDate: z.object({ date: z.unknown() }).nullable(),
});

const responseSchema = z.object({
  organization: z
    .object({
      projectV2: z
        .object({
          title: z.string(),
          items: z.object({
            totalCount: z.int().nonnegative(),
            nodes: z.array(itemSchema),
            pageInfo: z.object({
              hasNextPage: z.boolean(),
              endCursor: z.string().nullable(),
            }),
          }),
        })
        .nullable(),
    })
    .nullable(),
});

const githubDateSchema = z.iso
  .date()
  .transform((value) => Temporal.PlainDate.from(value));

const envSchema = z.object({ PROJECT_TOKEN: z.string().min(1) });

export type ProjectItem = z.infer<typeof itemSchema>;
export type GraphqlClient = (
  query: string,
  variables: Record<string, unknown>,
) => Promise<unknown>;

export interface CalendarEvent {
  uid: string;
  title: string;
  url: string;
  start: Temporal.PlainDate;
  end: Temporal.PlainDate;
  updatedAt: Temporal.Instant;
}

export async function fetchProject(client: GraphqlClient): Promise<{
  title: string;
  items: ProjectItem[];
}> {
  let cursor: string | null = null;
  let title = "calendar";
  let expectedCount: number | null = null;
  const items: ProjectItem[] = [];

  do {
    const response = responseSchema.parse(
      await client(query, {
        organization: ORGANIZATION,
        project: PROJECT_NUMBER,
        startField: START_FIELD,
        endField: END_FIELD,
        cursor,
      }),
    );
    const project = response.organization?.projectV2;
    if (!project) {
      throw new Error(
        `Project ${ORGANIZATION}#${PROJECT_NUMBER} was not found or is not accessible`,
      );
    }

    title = project.title;
    expectedCount = project.items.totalCount;
    items.push(...project.items.nodes);
    const pageInfo = project.items.pageInfo;
    if (pageInfo.hasNextPage && !pageInfo.endCursor) {
      throw new Error("GitHub returned a page without the required end cursor");
    }
    cursor = pageInfo.hasNextPage ? pageInfo.endCursor : null;
  } while (cursor);

  if (items.length !== expectedCount) {
    throw new Error(
      `Expected ${expectedCount} project items, received ${items.length}`,
    );
  }
  return { title, items };
}

export function toCalendarEvent(item: ProjectItem): CalendarEvent | null {
  const issue = item.content;
  if (
    !issue ||
    issue.__typename !== "Issue" ||
    issue.repository.nameWithOwner !== REPOSITORY
  ) {
    return null;
  }

  const parsedStart = githubDateSchema.safeParse(item.startDate?.date);
  const parsedEnd = githubDateSchema.safeParse(item.endDate?.date);
  let start = parsedStart.success ? parsedStart.data : null;
  let end = parsedEnd.success ? parsedEnd.data : null;
  if (!start && !end) return null;
  start ??= end;
  end ??= start;
  if (!start || !end) return null;

  if (Temporal.PlainDate.compare(end, start) < 0) {
    console.warn(`Skipped ${issue.url}: ${END_FIELD} precedes ${START_FIELD}`);
    return null;
  }

  return {
    uid: `${REPOSITORY}#${issue.number}@github.com`,
    title: issue.title,
    url: issue.url,
    start,
    end,
    updatedAt: Temporal.Instant.from(item.updatedAt),
  };
}

export function renderCalendar(name: string, events: CalendarEvent[]): string {
  const calendar = ical({
    name,
    prodId: {
      company: "zunoser",
      product: "GitHub Project Calendar",
      language: "EN",
    },
  });
  calendar.method(ICalCalendarMethod.PUBLISH);

  for (const event of events.toSorted((left, right) => {
    return (
      Temporal.PlainDate.compare(left.start, right.start) ||
      Temporal.PlainDate.compare(left.end, right.end) ||
      left.uid.localeCompare(right.uid)
    );
  })) {
    calendar.createEvent({
      id: event.uid,
      start: event.start,
      end: event.end.add({ days: 1 }),
      allDay: true,
      summary: event.title,
      url: event.url,
      stamp: event.updatedAt,
      lastModified: event.updatedAt,
    });
  }
  return calendar.toString();
}

export async function main(): Promise<void> {
  const { PROJECT_TOKEN } = envSchema.parse(process.env);
  const request = graphql.defaults({
    headers: { authorization: `token ${PROJECT_TOKEN}` },
  });
  const { title, items } = await fetchProject(request);
  const events = items
    .map(toCalendarEvent)
    .filter((event): event is CalendarEvent => event !== null);
  await Bun.write("calendar.ics", renderCalendar(title, events));
}

if (import.meta.main) {
  await main();
}
